fendermint/eth/api/src/apis/eth.rs

```diff
@@ -81,6 +89,121 @@ where
+    // for all blocks)
+    let consensus_params: consensus_params::Response = data
+        .tm()
+        .consensus_params(latest_h)
+        .await
+        .context("failed to get consensus params")?;
+    let mut block_gas_limit = consensus_params.consensus_params.block.max_gas;
+    if block_gas_limit <= 0 {
+        block_gas_limit =
+            i64::try_from(fvm_shared::BLOCK_GAS_LIMIT).expect("FVM block gas limit not i64")
+    };
+
+    let mut premiums = Vec::new();
+    // iterate through the blocks in the range
+    // we may be able to de-duplicate a lot of this code from fee_history
+    let latest_h: u64 = latest_h.into();
+    let mut blk = latest_h;
+    while blk > latest_h - NUM_BLOCKS_MAX_PRIO_FEE {
+        let block = data
+            .block_by_height(blk.into())
+            .await
+            .context("failed to get block")?;
+
+        let height = block.header().height;
+
+        // Genesis has height 1, but no relevant fees.
+        if height.value() <= 1 {
+            break;
+        }
+        let state_params = data.client.state_params(Some(height)).await?;
+        let base_fee = &state_params.value.base_fee;
+
+        // The latest block might not have results yet.
+        if let Ok(block_results) = data.tm().block_results(height).await {
+            let txs_results = block_results.txs_results.unwrap_or_default();
+
+            for (tx, txres) in block.data().iter().zip(txs_results) {
+                let msg = fvm_ipld_encoding::from_slice::<ChainMessage>(tx)
+                    .context("failed to decode tx as ChainMessage")?;
+
+                if let ChainMessage::Signed(msg) = msg {
+                    let premium = crate::gas::effective_gas_premium(&msg.message, base_fee);
+                    premiums.push((premium, txres.gas_used));
+                }
+            }
+        }
+        blk -= 1;
+    }
+
+    // compute median gas price
+    let mut median = median_gas_premium(&mut premiums, block_gas_limit);
+    let min_premium = TokenAmount::from_atto(BigInt::from(MIN_GAS_PREMIUM));
+    if median < min_premium {
+        median = min_premium;
+    }
+
+    // add some noise to normalize behaviour of message selection
+    // mean 1, stddev 0.005 => 95% within +-1%
+    const PRECISION: u32 = 32;
+    let mut rng = rand::thread_rng();
+    let noise: f64 = 1.0 + rng.gen::<f64>() * 0.005;
+    let precision: i64 = 32;
+    let coeff: u64 = ((noise * (1 << precision) as f64) as u64) + 1;
+
+    median *= BigInt::from(coeff);
+    median.div_ceil(BigInt::from(1 << PRECISION));
+
+    Ok(to_eth_tokens(&median)?)
+}
+
+// finds 55th percntile instead of median to put negative pressure on gas price
+// Rust implementation of:
+// https://github.com/consensus-shipyard/lotus/blob/156f5556b3ecc042764d76308dca357da3adfb4d/node/impl/full/gas.go#L144
+fn median_gas_premium(prices: &mut Vec<(TokenAmount, i64)>, block_gas_target: i64) -> TokenAmount {
```

Comment:
Could you move this helper function into the [gas module](https://github.com/consensus-shipyard/fendermint/blob/main/fendermint/eth/api/src/gas/mod.rs) please? 

I wanted to keep the `eth.rs` file free of any helpers, so every method in here corresponds to an API endpoint, so we know those are here, and everything else is not here.

Reply:
Sure, I didn't see the module there, that makes sense!